File: leetcode/grind75/implement-queue-using-stacks.ts
class MyStack {
  stack: Array<number>;
  
  constructor() {
      this.stack = [];
  }
  
  push(x: number): void {
      this.stack.push(x);
  }

  peek(): number {
      return this.stack[this.stack.length - 1];
  }

  pop(): number {
      return this.stack.pop();
  }

  size(): number {
      return this.stack.length;
  }

  isEmpty(): boolean {
      return this.stack.length === 0;
  }
}

// input: [1, 2, 3, 4, 5]
// push input in, pop input out
// stack always pop out of back
// top of stack is on the right
// stack1: [4, 3, 2, 1]
// stack2: []


class MyQueue {
  stack1: MyStack;
  stack2: MyStack;

  constructor() {
      this.stack1 = new MyStack();
      this.stack2 = new MyStack();
  }

  push(x: number): void {
      // pop everything into stack 2
      while (!this.stack1.isEmpty()) {
          this.stack2.push(this.stack1.pop());
      }
      
      // put new number into stack 1
      this.stack1.push(x);

      // pop everything back from stack 2 into stack 1
      while (!this.stack2.isEmpty()) {
          this.stack1.push(this.stack2.pop());
      }
  }

  pop(): number {
      return this.stack1.pop();
  }

  peek(): number {
      return this.stack1.peek();
  }

  empty(): boolean {
      return this.stack1.isEmpty();
  }
}

/**
* Your MyQueue object will be instantiated and called as such:
* var obj = new MyQueue()
* obj.push(x)
* var param_2 = obj.pop()
* var param_3 = obj.peek()
* var param_4 = obj.empty()
*/

File: leetcode/grind75/implement-queue-using-stacks.rs
impl MyStack {
  fn new() -> Self {
      MyStack { stack: vec![] }
  }

  fn push(&mut self, x: i32) {
      self.stack.push(x);
  }

  fn peek(&self) -> i32 {
      self.stack[self.stack.len() - 1]
  }

  fn pop(&mut self) -> i32 {
      self.stack.pop().unwrap()
  }

  fn size(&self) -> usize {
      self.stack.len()
  }

  fn is_empty(&self) -> bool {
      self.stack.len() == 0
  }
}

struct MyQueue {
  main_stack: MyStack,
  transfer_stack: MyStack,
}

/**
* `&self` means the method takes an immutable reference.
* If you need a mutable reference, change it to `&mut self` instead.
*/
impl MyQueue {
  fn new() -> Self {
      MyQueue {
          main_stack: MyStack::new(),
          transfer_stack: MyStack::new(),
      }
  }

  fn push(&mut self, x: i32) {
      // Pop everything from main stack and push into transfer stack
      while !self.main_stack.is_empty() {
          self.transfer_stack.push(self.main_stack.pop());
      }

      // Push x into main stack
      self.main_stack.push(x);

      // Pop everything back from transfer to main stack
      while !self.transfer_stack.is_empty() {
          self.main_stack.push(self.transfer_stack.pop());
      }
  }

  fn pop(&mut self) -> i32 {
      self.main_stack.pop()
  }

  fn peek(&self) -> i32 {
      self.main_stack.peek()
  }

  fn empty(&self) -> bool {
      self.main_stack.is_empty()
  }
}

/**
* Your MyQueue object will be instantiated and called as such:
* let obj = MyQueue::new();
* obj.push(x);
* let ret_2: i32 = obj.pop();
* let ret_3: i32 = obj.peek();
* let ret_4: bool = obj.empty();
*/